docs/adr/0009-uniffi-tagging.md

```diff
@@ -0,0 +1,95 @@
+This is quite confusing in the abstract, hopefully the example below clears things up.
+
+Currently, UDL-based generation uses the former approach and proc-macros use the latter.  We should pick one approach and stick with it.
+
+### Example
+
+- `crate_a` depends on the `bitcoin` crate and uses the [Network](https://docs.rs/bitcoin/latest/bitcoin/network/enum.Network.html) enum in their exported API.
+  (Let's assume that we've implemented a way to do this with proc-macros, maybe some system similar to [the way serde handles it](https://serde.rs/remote-derive.html))
+- `crate_a` defines/exports a `NFTGenerator` type, which is just a normal UniFFI interface.
+- `crate_b` depends on `crate_a` and `bitcoin` and wants to define the function `fn build_nft_generator(network: Network) -> NFTGenerator`.
+  This is a convenience function that simplifies constructing an `NFTGenerator`.
+
+When `crate_a` implement the FFI traits for `Network` it must implement them for its local `UniffiTag` only.
+However, it can implement the FFI traits for `NFTGenerator` for either its local `UniffiTag` or all tags.
+This choice will affect how `crate_b` "imports" the UniFFI types.
+
+### Always implement traits for the local tag only
+
+- `crate_a` implements the FFI traits for its local `UniffiTag` for both `Network` and `NFTGenerator`.
+- `crate_b` needs to take explicit action to use the UniFFI types from `crate_a`.
+
+```rust
+
+use crate_a::{NFTGenerator, Network};
+
+uniffi::use_object!(crate_a, NFTGenerator)
+uniffi::use_enum!(crate_a, Network)
+```
+
+Note: right now these macros are named `use_udl_object`, etc.
+Once proc-macros can implement traits for remote types we probably should rename them to something not specific to UDL, but the exact name is out of scope for this ADR.
+
+### Implement traits for all tags when possible
+
+
+- `crate_a` implements the FFI traits for its local `UniffiTag` for `Network`
+- `crate_a` implements the FFI traits for all tags for `NFTGenerator`.
+  This is allowed since the type is local to `crate_a`.
+- `crate_b` only needs to take explicit action to use `Network` since it's a remote type for `crate_a`.
+
+```rust
+
+use crate_a::{NFTGenerator, Network};
+
+uniffi::use_enum!(crate_a, Network)
+```
+
+## Decision Drivers
+
+- Using one system for UDL and a different one for proc-macros is overly complicated and therefore not discussed as an option.
+- TODO: add more as we discuss this.
+
+## Considered Options
+
+### [Option 1] Always implement FFI traits for the local tag only
+
+### [Option 2] Implement FFI traits for all tags for non-remote tags
+
+## Pros and Cons of the Options
+
+### [Option 1] Always implement FFI traits for the local tag only
+
+- Good, because it's more explicit.
+- Good, because it's easier to explain.
+  Telling users "you always have to add `uniffi::use_*!`" to use a type from another crate is easier than saying you sometimes you need it and sometimes you don't.
```

Comment:
This factor seems the most important to me.   I don't want to have to go into the details about the orphan rule, remote types, etc. in order to explain to users if they need `uniffi::use_record!` or not.